tests/0151-purge-brokers.c

```diff
@@ -0,0 +1,410 @@
+ *         and \p actual_broker_ids correspond in
+ *         count and value. Sorts \p actual_broker_ids .
+ */
+static rd_bool_t fetch_metadata_verify_brokers(int32_t *expected_broker_ids,
+                                               size_t expected_broker_id_cnt,
+                                               int32_t *actual_broker_ids,
+                                               size_t actual_broker_id_cnt) {
+        size_t i;
+        if (actual_broker_id_cnt != expected_broker_id_cnt)
+                return rd_false;
+
+        for (i = 0; i < actual_broker_id_cnt; i++) {
+                if (actual_broker_ids[i] != expected_broker_ids[i])
+                        return rd_false;
+        }
+        return rd_true;
+}
+
+static void fetch_metadata(rd_kafka_t *rk,
+                           int32_t *expected_broker_ids,
+                           size_t expected_broker_id_cnt,
+                           rd_bool_t (*await_after_action_cb)(int action),
+                           int action) {
+        const rd_kafka_metadata_t *md = NULL;
+        rd_kafka_resp_err_t err;
+        size_t actual_broker_id_cnt = 0;
+        int32_t *actual_broker_ids  = NULL;
+        size_t i;
+        int timeout_usecs                      = 10000000;
+        int64_t abs_timeout_us                 = test_clock() + timeout_usecs;
+        rd_bool_t continue_requesting_metadata = rd_true;
+
+        TEST_SAY("Waiting for up to 10s for metadata update\n");
+
+        /* Trigger Metadata request which will update learned brokers. */
+        do {
+                err = rd_kafka_metadata(rk, 0, NULL, &md, tmout_multip(5000));
+                if (md) {
+                        rd_kafka_metadata_destroy(md);
+                        md = NULL;
+                } else if (err != RD_KAFKA_RESP_ERR__TRANSPORT)
+                        TEST_ASSERT(!err, "%s", rd_kafka_err2str(err));
+
+                rd_usleep(100 * 1000, 0);
+
+                RD_IF_FREE(actual_broker_ids, rd_free);
+                actual_broker_ids =
+                    rd_kafka_brokers_learned_ids(rk, &actual_broker_id_cnt);
+                continue_requesting_metadata = test_clock() <= abs_timeout_us;
+
+                continue_requesting_metadata &= !fetch_metadata_verify_brokers(
+                    expected_broker_ids, expected_broker_id_cnt,
+                    actual_broker_ids, actual_broker_id_cnt);
+
+                if (await_after_action_cb)
+                        continue_requesting_metadata |=
+                            await_after_action_cb(action);
+
+        } while (continue_requesting_metadata);
+
+        TEST_ASSERT(actual_broker_id_cnt == expected_broker_id_cnt,
+                    "expected %" PRIusz " brokers in cache, got %" PRIusz,
+                    expected_broker_id_cnt, actual_broker_id_cnt);
+
+        for (i = 0; i < actual_broker_id_cnt; i++) {
+                TEST_ASSERT(actual_broker_ids[i] == expected_broker_ids[i],
```

Comment:
Better to sort `expected_broker_ids` in the function instead of expecting it to be sorted.

Reply:
There's the change to return `rd_kafka_brokers_learned_ids` from `rk_broker_by_id` that is sorted. So it's ensured. I can add it to the function documentation.

Reply:
Yes but the list of `expected_broker_ids` is defined by the user which might not be sorted. Its better to sort it.